presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -24,41 +25,108 @@
+import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.toMap;
 
 public class ClusterManager
 {
-    private final Map<String, GroupSpec> groups;
-    private final List<SelectorRuleSpec> groupSelectors;
-    private final SchedulerType schedulerType;
-    private final Scheduler scheduler;
-    private final HashMap<String, HashMap<URI, Integer>> serverWeights = new HashMap<>();
+    private Map<String, GroupSpec> groups;
+    private List<SelectorRuleSpec> groupSelectors;
+    private SchedulerType schedulerType;
+    private Scheduler scheduler;
+    private HashMap<String, HashMap<URI, Integer>> serverWeights = new HashMap<>();
+    private HashMap<URI, URI> discoveryURIs = new HashMap<>();
+    private final RouterConfig routerConfig;
+    private final ScheduledExecutorService scheduledExecutorService;
+    private final AtomicLong lastConfigUpdate = new AtomicLong();
+    private final RemoteInfoFactory remoteInfoFactory;
+    private final Logger log = Logger.get(ClusterManager.class);
+
+    // Cluster status
+    private final ConcurrentHashMap<URI, RemoteClusterInfo> remoteClusterInfos = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<URI, RemoteQueryInfo> remoteQueryInfos = new ConcurrentHashMap<>();
 
     @Inject
-    public ClusterManager(RouterConfig config)
+    public ClusterManager(RouterConfig config, @ForClusterManager ScheduledExecutorService scheduledExecutorService, RemoteInfoFactory remoteInfoFactory)
     {
+        this.routerConfig = config;
+        this.scheduledExecutorService = scheduledExecutorService;
         RouterSpec routerSpec = parseRouterConfig(config)
                 .orElseThrow(() -> new PrestoException(CONFIGURATION_INVALID, "Failed to load router config"));
-
         this.groups = ImmutableMap.copyOf(routerSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
         this.groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
         this.schedulerType = routerSpec.getSchedulerType();
         this.scheduler = new SchedulerFactory(routerSpec.getSchedulerType()).create();
-
+        this.remoteInfoFactory = requireNonNull(remoteInfoFactory, "remoteInfoFactory is null");
         this.initializeServerWeights();
+        this.initializeMembersDiscoveryURI();
+        List<URI> allClusters = getAllClusters();
+        allClusters.forEach(uri -> {
+            log.info("Attaching cluster %s to the router", uri.getHost());
+            remoteClusterInfos.put(uri, remoteInfoFactory.createRemoteClusterInfo(discoveryURIs.get(uri)));
+            remoteQueryInfos.put(uri, remoteInfoFactory.createRemoteQueryInfo(discoveryURIs.get(uri)));
+            log.info("Successfully attached cluster %s to the router. Queries will be routed to cluster after successful health check", uri.getHost());
+        });
+    }
+
+    @PostConstruct
+    public void startConfigReloadTask()
```

Comment:
Can you add a test for this scenario - the file config gets updated, old servers are removed. New ones get added, existing ones stay as-is